File: dart/test/semantic_index_query_kernel_test.dart
// FUTURE-PARITY-BACKLOG.10.5.4.1 — typed record/source query kernel.

import 'dart:convert';
import 'dart:io';

import 'package:linkedspec_dart/src/semantic/semantic_index.dart';
import 'package:linkedspec_dart/src/semantic/sha256.dart';
import 'package:test/test.dart';

void main() {
  test('typed record source kernel matches nine exact neutral digests', () {
    for (final id in [
      'capabilities',
      'graph_list_rules',
      'graph_duplicate_regex_text',
      'graph_explain_entry',
      'calls_symbols_and_shapes',
      'failed_diagnostic',
      'privacy_none',
      'privacy_text_and_digest',
      'source_ceiling_forbidden',
    ]) {
      final queryCase = _queryCase(id);
      final request = _typedRequest(
        queryCase['request']! as Map<String, Object?>,
      );
      final index = _indexFor(queryCase['snapshot']! as String);
      final response = id == 'capabilities'
          ? index.semanticCapabilitiesForTesting()
          : index.semanticQueryKernelForTesting(request);
      final expected = queryCase['expected']! as Map<String, Object?>;

      expect(response.ok, expected['ok'], reason: '$id status');
      expect(
        response.records.map((record) => record.id),
        expected['record_ids'],
        reason: '$id records',
      );
      expect(
        response.relations.map((relation) => relation.id),
        expected['relation_ids'],
        reason: '$id relations',
      );
      expect(
        response.diagnostics.map((diagnostic) => diagnostic.code),
        expected['diagnostic_codes'],
        reason: '$id diagnostics',
      );
      expect(response.page.complete, expected['complete'], reason: id);
      expect(
        _canonicalDigest(response.toJson()),
        expected['response_sha256'],
        reason: '$id full response digest',
      );
    }
  });

  test('capabilities and response values are immutable fresh clones', () {
    final queryCase = _queryCase('capabilities');
    final expected =
        (queryCase['expected']! as Map<String, Object?>)['response_sha256'];
    final index = _indexFor('graph');
    final first = index.semanticCapabilitiesForTesting();

    expect(
      () => first.records.add(first.records.single),
      throwsUnsupportedError,
    );
    expect(
      () => first.records.single.facts['model_ids'] = ['host-private'],
      throwsUnsupportedError,
    );
    final detached = first.toJson();
    final records = detached['records']! as List<Object?>;
    final capability = records.single as Map<String, Object?>;
    final facts = capability['facts']! as Map<String, Object?>;
    (facts['record_kinds']! as List<Object?>)[0] = 'host-private';

    final second = index.semanticCapabilitiesForTesting();
    expect(second, isNot(same(first)));
    expect(_canonicalDigest(second.toJson()), expected);
    expect(index.compilationAuthority.compiled, isTrue);
    expect(index.snapshot.hasExecution, isFalse);
  });

  test('source privacy and structural redaction are monotonic', () {
    final privacy = _indexFor('privacy');
    final none = privacy.semanticQueryKernelForTesting(
      _typedRequest(_requestFor('privacy_none')),
    );
    expect(none.records.single.source, isNull);
    expect(none.records.single.facts['pattern'], isNull);
    expect(none.records.single.redactions, ['/facts/pattern']);

    final text = privacy.semanticQueryKernelForTesting(
      _typedRequest(_requestFor('privacy_text_and_digest')),
    );
    expect(text.records.single.facts['pattern'], 'é');
    expect(text.records.single.redactions, isEmpty);
    expect(text.records.single.source?.excerpt, '/é/');
    expect(
      text.records.single.source?.contentDigest,
      allOf(startsWith('sha256:'), hasLength(71)),
    );

    final limited = _indexFor('privacy_limited').semanticQueryKernelForTesting(
      _typedRequest(_requestFor('source_ceiling_forbidden')),
    );
    expect(limited.ok, isFalse);
    expect(limited.diagnostics.single.fields, {
      'requested': 'span',
      'ceiling': 'identity',
    });
  });

  test('kernel is deterministic projection-only and not publicly exported', () {
    final index = _indexFor('graph');
    final explain = _typedRequest(_requestFor('graph_explain_entry'));
    final list = _typedRequest(_requestFor('graph_list_rules'));
    final first = index.semanticQueryKernelForTesting(explain);
    final middle = index.semanticQueryKernelForTesting(list);
    final second = index.semanticQueryKernelForTesting(explain);

    expect(first, second);
    expect(first, isNot(middle));
    final encoded = jsonEncode(first.toJson());
    for (final forbidden in [
      '/tmp/',
      '/Users/',
      'CompiledSpec',
      'ActionIR',
      'Regex(',
      '0x',
      'generated source',
    ]) {
      expect(encoded, isNot(contains(forbidden)), reason: forbidden);
    }

    final implementation = File(
      'lib/src/semantic/semantic_query.dart',
    ).readAsStringSync();
    for (final forbidden in [
      '_sourceText',
      '_sourceBytes',
      '_sourceMap',
      '_compilationOutcome',
      'parseSpecWithStagedUserFunctionDefinitions(',
      'compileSpec(',
      'emitDartSource(',
      'executeGeneratedParser',
      'LinkedSpecRuntimeEngine',
      'Platform.environment',
      'File(',
    ]) {
      expect(implementation, isNot(contains(forbidden)), reason: forbidden);
    }

    final publicLibrary = File('lib/linkedspec_dart.dart').readAsStringSync();
    for (final omitted in [
      'SemanticQueryOperation',
      'SemanticQueryResponse',
      'SemanticIndexQueryKernelTestAccess',
    ]) {
      expect(publicLibrary, isNot(contains(omitted)), reason: omitted);
    }
  });
}

final Map<String, Object?> _contract =
    jsonDecode(
          File(
            '../capability_conformance/semantic_introspection_contract.json',
          ).readAsStringSync(),
        )!
        as Map<String, Object?>;

Map<String, Object?> _queryCase(String id) =>
    (_contract['query_cases']! as List<Object?>)
        .cast<Map<String, Object?>>()
        .singleWhere((queryCase) => queryCase['id'] == id);

Map<String, Object?> _requestFor(String id) =>
    _queryCase(id)['request']! as Map<String, Object?>;

SemanticIndex _indexFor(String snapshot) {
  final (fixture, ceiling) = switch (snapshot) {
    'graph' => ('graph', SemanticSourceDetail.text),
    'calls' => ('calls_and_staging', SemanticSourceDetail.text),
    'failed' => ('failed', SemanticSourceDetail.span),
    'privacy' => ('privacy', SemanticSourceDetail.text),
    'privacy_limited' => ('privacy', SemanticSourceDetail.identity),
    _ => throw StateError('Unexpected static snapshot: $snapshot'),
  };
  return SemanticIndex.fromUtf8(
    File(
      '../capability_conformance/semantic_introspection/$fixture.spec',
    ).readAsBytesSync(),
    options: SemanticIndexOptions(
      logicalName: '$fixture.spec',
      sourceDetailCeiling: ceiling,
    ),
  );
}

SemanticQuery _typedRequest(Map<String, Object?> value) {
  final page = value['page']! as Map<String, Object?>;
  final budget = value['budget']! as Map<String, Object?>;
  final source = value['source']! as Map<String, Object?>;
  return SemanticQuery(
    contract: value['contract']! as String,
    operation: switch (value['operation']) {
      'capabilities' => SemanticQueryOperation.capabilities,
      'list' => SemanticQueryOperation.list,
      'get' => SemanticQueryOperation.get,
      'relations' => SemanticQueryOperation.relations,
      'explain' => SemanticQueryOperation.explain,
      final operation => throw StateError('Unexpected operation: $operation'),
    },
    subjects: (value['subjects']! as List<Object?>).cast<String>(),
    recordKinds: (value['record_kinds']! as List<Object?>).cast<String>(),
    relationKinds: (value['relation_kinds']! as List<Object?>).cast<String>(),
    direction: switch (value['direction']) {
      'outgoing' => SemanticQueryDirection.outgoing,
      'incoming' => SemanticQueryDirection.incoming,
      'both' => SemanticQueryDirection.both,
      final direction => throw StateError('Unexpected direction: $direction'),
    },
    page: SemanticQueryPage(
      afterId: page['after_id'] as String?,
      limit: page['limit']! as int,
    ),
    budget: SemanticQueryBudget(
      maxRecords: budget['max_records']! as int,
      maxRelations: budget['max_relations']! as int,
      maxDepth: budget['max_depth']! as int,
    ),
    source: SemanticQuerySource(
      detail: switch (source['detail']) {
        'none' => SemanticSourceDetail.none,
        'identity' => SemanticSourceDetail.identity,
        'span' => SemanticSourceDetail.span,
        'text' => SemanticSourceDetail.text,
        final detail => throw StateError('Unexpected source detail: $detail'),
      },
      includeContentDigest: source['include_content_digest']! as bool,
    ),
  );
}

String _canonicalDigest(Map<String, Object?> value) =>
    sha256Hex(utf8.encode(jsonEncode(_canonicalValue(value))));

Object? _canonicalValue(Object? value) {
  if (value case Map<Object?, Object?>()) {
    final keys = value.keys.map((key) => key.toString()).toList()..sort();
    return <String, Object?>{
      for (final key in keys) key: _canonicalValue(value[key]),
    };
  }
  if (value case List<Object?>()) {
    return [for (final item in value) _canonicalValue(item)];
  }
  return value;
}

File: julia/test/semantic_index_query_kernel_test.jl
# FUTURE-PARITY-BACKLOG.10.6.5.1 — private immutable non-traversal query kernel.

const _SEMANTIC_QUERY_KERNEL_CONTRACT = JSON3.read(
    read(
        joinpath(
            REPO_ROOT,
            "capability_conformance",
            "semantic_introspection_contract.json",
        ),
        String,
    ),
    Dict{String,Any},
)

const _SEMANTIC_QUERY_KERNEL_CASES = (
    "capabilities",
    "graph_list_rules",
    "graph_duplicate_regex_text",
    "graph_explain_entry",
    "calls_symbols_and_shapes",
    "failed_diagnostic",
    "privacy_none",
    "privacy_text_and_digest",
    "source_ceiling_forbidden",
)

function _semantic_query_kernel_case(id)
    return only(
        query_case for query_case in _SEMANTIC_QUERY_KERNEL_CONTRACT["query_cases"] if
        query_case["id"] == id
    )
end

function _semantic_query_kernel_index(snapshot)
    fixture, logical_name, ceiling = if snapshot == "graph"
        ("graph.spec", "graph.spec", SemanticSourceTextDetail)
    elseif snapshot == "calls"
        ("calls_and_staging.spec", "calls_and_staging.spec", SemanticSourceTextDetail)
    elseif snapshot == "failed"
        ("failed.spec", "failed.spec", SemanticSourceSpanDetail)
    elseif snapshot == "privacy"
        ("privacy.spec", "privacy.spec", SemanticSourceTextDetail)
    elseif snapshot == "privacy_limited"
        ("privacy.spec", "privacy.spec", SemanticSourceIdentityDetail)
    else
        error("Unknown semantic query kernel snapshot: $snapshot")
    end
    source = read(
        joinpath(
            REPO_ROOT,
            "capability_conformance",
            "semantic_introspection",
            fixture,
        ),
        String,
    )
    return semantic_index(
        source;
        logical_name = logical_name,
        source_detail_ceiling = ceiling,
    )
end

function _semantic_query_kernel_request(value)
    operation = Dict(
        "capabilities" => LinkedSpecJulia.SemanticQueryCapabilitiesOperation,
        "list" => LinkedSpecJulia.SemanticQueryListOperation,
        "get" => LinkedSpecJulia.SemanticQueryGetOperation,
        "relations" => LinkedSpecJulia.SemanticQueryRelationsOperation,
        "explain" => LinkedSpecJulia.SemanticQueryExplainOperation,
    )[value["operation"]]
    direction = Dict(
        "outgoing" => LinkedSpecJulia.SemanticQueryOutgoingDirection,
        "incoming" => LinkedSpecJulia.SemanticQueryIncomingDirection,
        "both" => LinkedSpecJulia.SemanticQueryBothDirection,
    )[value["direction"]]
    detail = Dict(
        "none" => SemanticSourceNoneDetail,
        "identity" => SemanticSourceIdentityDetail,
        "span" => SemanticSourceSpanDetail,
        "text" => SemanticSourceTextDetail,
    )[value["source"]["detail"]]
    return LinkedSpecJulia.SemanticQuery(
        operation;
        contract = value["contract"],
        subjects = value["subjects"],
        record_kinds = value["record_kinds"],
        relation_kinds = value["relation_kinds"],
        direction = direction,
        page = LinkedSpecJulia.SemanticQueryPage(
            after_id = value["page"]["after_id"],
            limit = value["page"]["limit"],
        ),
        budget = LinkedSpecJulia.SemanticQueryBudget(
            max_records = value["budget"]["max_records"],
            max_relations = value["budget"]["max_relations"],
            max_depth = value["budget"]["max_depth"],
        ),
        source = LinkedSpecJulia.SemanticQuerySource(
            detail = detail,
            include_content_digest = value["source"]["include_content_digest"],
        ),
    )
end

function _semantic_query_kernel_digest(response)
    encoded = LinkedSpecJulia._primary_cli_canonical_json(to_json(response))
    return bytes2hex(LinkedSpecJulia.SHA.sha256(codeunits(encoded)))
end

function _semantic_query_kernel_has_no_mutable_container(value)
    if value isa AbstractDict || value isa AbstractVector
        return false
    elseif value isa LinkedSpecJulia._SemanticQueryObject ||
           value isa LinkedSpecJulia._SemanticQueryArray
        return all(
            item -> item isa Pair ?
                    _semantic_query_kernel_has_no_mutable_container(last(item)) :
                    _semantic_query_kernel_has_no_mutable_container(item),
            getfield(value, :values),
        )
    elseif value isa Tuple
        return all(_semantic_query_kernel_has_no_mutable_container, value)
    end
    return true
end

@testset "Semantic index private immutable query kernel" begin
    responses = Dict{String,Any}()

    @testset "nine exact non-traversal response hashes" begin
        @test length(_SEMANTIC_QUERY_KERNEL_CASES) == 9
        for id in _SEMANTIC_QUERY_KERNEL_CASES
            query_case = _semantic_query_kernel_case(id)
            request = _semantic_query_kernel_request(query_case["request"])
            index = _semantic_query_kernel_index(query_case["snapshot"])
            response = LinkedSpecJulia._semantic_query_kernel(index, request)
            expected = query_case["expected"]
            responses[id] = response

            @test response.ok == expected["ok"]
            @test [record.id for record in response.records] == expected["record_ids"]
            @test [relation.id for relation in response.relations] == expected["relation_ids"]
            @test [diagnostic.code for diagnostic in response.diagnostics] ==
                  expected["diagnostic_codes"]
            @test response.page.complete == expected["complete"]
            @test _semantic_query_kernel_digest(response) == expected["response_sha256"]
        end
    end

    @testset "source privacy and exact structural redaction" begin
        none = to_json(responses["privacy_none"])["records"][1]
        @test none["source"] === nothing
        @test none["facts"]["pattern"] === nothing
        @test none["redactions"] == Any["/facts/pattern"]

        text = to_json(responses["privacy_text_and_digest"])["records"][1]
        @test text["facts"]["pattern"] == "é"
        @test text["redactions"] == Any[]
        @test text["source"]["excerpt"] == "/é/"
        @test startswith(text["source"]["content_digest"], "sha256:")
        @test ncodeunits(text["source"]["content_digest"]) == 71

        limited = responses["source_ceiling_forbidden"]
        @test !limited.ok
        @test to_json(only(limited.diagnostics))["fields"] == Dict{String,Any}(
            "requested" => "span",
            "ceiling" => "identity",
        )
    end

    @testset "recursive immutability and fresh detached serialization" begin
        index = _semantic_query_kernel_index("graph")
        request = _semantic_query_kernel_request(
            _semantic_query_kernel_case("capabilities")["request"],
        )
        first = LinkedSpecJulia._semantic_query_kernel(index, request)
        first_json = to_json(first)
        first_json["records"][1]["facts"]["record_kinds"][1] = "host-private"
        second = LinkedSpecJulia._semantic_query_kernel(index, request)

        @test first == second
        @test to_json(first) !== to_json(second)
        @test to_json(first)["records"] !== to_json(second)["records"]
        @test _semantic_query_kernel_digest(second) ==
              _semantic_query_kernel_case("capabilities")["expected"]["response_sha256"]
        @test to_json(first)["records"][1]["facts"]["record_kinds"][1] == "capabilities"
        @test all(
            record ->
                _semantic_query_kernel_has_no_mutable_container(record.facts) &&
                _semantic_query_kernel_has_no_mutable_container(record.redactions),
            first.records,
        )
        @test getfield(index, :_static_projection) isa LinkedSpecJulia._SemanticStaticProjection
    end

    @testset "typed boundaries and private kernel seam" begin
        graph = _semantic_query_kernel_index("graph")
        relations = LinkedSpecJulia.SemanticQuery(
            LinkedSpecJulia.SemanticQueryRelationsOperation;
            subjects = ("rule:Top",),
        )
        paged = LinkedSpecJulia.SemanticQuery(
            LinkedSpecJulia.SemanticQueryListOperation;
            page = LinkedSpecJulia.SemanticQueryPage(limit = 1),
        )
        budgeted = LinkedSpecJulia.SemanticQuery(
            LinkedSpecJulia.SemanticQueryListOperation;
            budget = LinkedSpecJulia.SemanticQueryBudget(max_records = 2),
        )
        @test LinkedSpecJulia._semantic_query_kernel(graph, relations).ok
        @test LinkedSpecJulia._semantic_query_kernel(graph, paged).page.next_after_id !== nothing
        @test !LinkedSpecJulia._semantic_query_kernel(graph, budgeted).page.complete
        @test_throws ArgumentError LinkedSpecJulia.SemanticQueryPage(limit = true)
        @test_throws MethodError LinkedSpecJulia.SemanticQueryPage(nothing, true)
        @test_throws ArgumentError LinkedSpecJulia.SemanticQueryBudget(max_records = true)
        @test_throws MethodError LinkedSpecJulia.SemanticQueryBudget(true, 2_000, 4)
        @test_throws ArgumentError LinkedSpecJulia.SemanticQuerySource(
            include_content_digest = 1,
        )
        @test LinkedSpecJulia.SemanticQuery(
            operation = LinkedSpecJulia.SemanticQueryListOperation,
        ) == LinkedSpecJulia.SemanticQuery(LinkedSpecJulia.SemanticQueryListOperation)

        public_names = names(LinkedSpecJulia)
        for name in (
            :SemanticQuery,
            :SemanticQueryResponse,
            :semantic_capabilities,
            :semantic_query,
            :semantic_query_neutral,
        )
            @test name in public_names
        end
        @test !(:_semantic_query_kernel in public_names)
    end

    @testset "detached-projection-only implementation" begin
        implementation = read(
            joinpath(REPO_ROOT, "julia", "src", "semantic", "SemanticQuery.jl"),
            String,
        )
        @test length(collect(eachmatch(
            r"_semantic_static_projection_materialize\(index\)",
            implementation,
        ))) == 1
        for forbidden in (
            "getfield(index, :_source_text)",
            "getfield(index, :_source_map)",
            "getfield(index, :_compilation_outcome)",
            "parse_spec(",
            "compile_spec(",
            "build_generated_rule_plan(",
            "emit_julia_source(",
            "runtime_execute(",
            "LinkedSpecTraceEmitter(",
            "ENV[",
            "time_ns(",
            "rand(",
        )
            @test !occursin(forbidden, implementation)
        end
    end
end
